2019/accidental.slide

```diff
@@ -0,0 +1,214 @@
+
+
+* go.science — G+ community
+
+Initial discussions were on G+
+
+.image accidental/images/go.science.png 500 _
+
+
+* Initial designs
+
+Focused around:
+
+- matrix arithmetic and linear algebra
+- simplifying `[]float64` handling
+- statistical functions
+- graph analysis
+
+Progressed to include numerical optimisation functions
+
+Attracted additional committed contributors to packages at [[https://github.com/gonum/][github.com/gonum]]
+
+
+* Small consistent contributing community
+
+.image accidental/images/contributors.png
+
+The people who contribute generally do so because it's fun
+
+
+* Largely influenced by the early golang-nuts culture
+
+I recall Russ Cox discussing bootstrapping of the culture in a talk
+
+# I couldn't find this in any videos, so I contacted Russ to ask if he remembered.
+# He didn't, but wrote the following, which matched my recollection perfectly.
+
+"I do think it was important to engage with everyone especially at the start, and then as other voices got up to speed and were able to handle most questions, we were able to step back."
+
+This helped build a strong early community of people wanting to build the things that were missing at the time
+
+
+* International efforts
+
+For the majority of the project the four main contributors lived on separate continents
+
+The first time meeting each other f2f or by video chat was this year
+
+- Didn't hamper development because of a strong common culture and Go's design
+
+
+* Gonum - core
+
+.iframe https://godoc.org/gonum.org/v1/gonum 550 950
+
+
+* Gonum - plotting
+
+
+.iframe https://godoc.org/gonum.org/v1/plot 550 950
+.caption Originally [[https://github.com/gonum/plot/wiki/Plotinum-redesign][Plotinum]] started by Ethan Burns
+
+
+* Gonum design principles
+
+* Consistency of API
+
+Doing the same kind of thing in different settings should look approximately the same
+(this is not R or perl)
+
+- Types usable composably between packages with little friction
+
+* No surprises
+
+Behaviour should be clearly documented and magic should not enter the picture
+(R, I am looking at you)
+
+- Testing is all just pure "testing" package
+
+- Close to zero use of `interface{}` in exported code
+
+
+* panics versus errors
+
+"*Don't*panic*" get's into the [[https://go-proverbs.github.io/][go-proverbs.io]] list from [[https://golang.org/doc/effective_go.html#panic][Effective Go]] but was not in the [[https://www.youtube.com/watch?v=PAAkCSZUG1c][talk]].
+
+Effective Go says:
+*But*what*if*the*error*is*unrecoverable?*
+*Sometimes*the*program*simply*cannot*continue.*
+
+Gonum follows the principle that incorrect software in a scientific setting should not continue running
+
+In the world of bioinformatics, incorrect results can lead to significant harm ([[https://www.nytimes.com/2011/07/08/health/research/08genes.html][erroneous drug trial approval for example]])
+
+Otherwise, idiomatic Go code
+
+
+* Notable Gonum package clients
+
+- Gorgonia — machine learning for Go
```

Comment:
perhaps add links to the projects of this list? (when available)

Reply:
Done